site/source/tutorials/005_form_widgets/index.md

```diff
@@ -1,26 +1,159 @@
+				content: 'Save',
+				type: 'submit'
+			})
+		]);
+	}
+```
+
+At this point, the user interface for the form is available, but it doesn't do anything since we haven't specified any event handlers. In the [last](../004_user_interactions/) tutorial, we learned how to add event handlers to custom widgets by assigning a method to an event. When using pre-built widgets, however, we pass the handlers as properties. For example, we are going to need a way to handle each text fields `change` event. To do that, we provide the desired handler function as the `onChange` property that is passed to the widget. Update the `render` method once again:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:33-110 %}
+
+This form of the `render` method now does everything that we need it to: it creates the user interface and registers the event handlers that will update the application as the user enters information.
+
+The method starts by decomposing the properties down in to local constants. We still need to define those properties, so update the `WorkerFormProperties` interface to include them:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:9-19 %}
+
+There are three types of properties that we are using in this form. The `firstName`, `lastName` and `email` properties are going to set the values that are displayed in the form fields. The `firstNameInvalid`, `lastNameInvalid` and `emailInvalid` properties will be used to indicate if a field contains invalid data (e.g. an invalid email address in the `email` field). Finally, the `onChange`, `onBlur` and `onSubmit` properties expose the events that the `WorkerForm` widget can emit. To see how these different property types are used, let's examine the properties that are being passed into first `TextInput` widget:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:51-69 %}
+
+The first thing that we see is a `key` property. As mentioned before, a key is necessary whenever the same type of widget or virtual DOM element will be rendered by a widget. After that comes the `invalid` property. Setting this property to `true` will update the widget's appearance to indicate that the widget contains invalid data. Next comes the the `label`, `placeholder`, and `type` fields which map to the expected properties.
+
+The `value` property renders the value of the field that is passed into the widget via its properties. Notice that there is no code that manipulates this value within the widget. As parts a [reactive framework](https://en.wikipedia.org/wiki/Reactive_programming), Dojo 2 widgets don't normally update their own state. Rather, they inform their parents that a change has occurred via events or some other mechanism. The parent will then pass updated properties back into the widget after all of the changes have been processed. This allows Dojo 2 applications to centralize data and keep the entire application synchronized.
```

Comment:
"they inform their parents" widgets only have one true parent, right? Maybe ancestor widgets or parent widget instead of parents?